File: leet/find-first-and-last-position-of-element-in-sorted-array.ts
/**
 * https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/
 */

export function searchRange(nums: number[], target: number): number[] {
  const foundIndex = binarySearch(nums, target);
  if (foundIndex === -1) return [-1, -1];

  let foundStart = foundIndex;
  while (nums[foundStart] === nums[foundStart - 1]) {
    foundStart--;
  }

  let foundEnd = foundIndex;
  while (nums[foundEnd] === nums[foundEnd + 1]) {
    foundEnd++;
  }

  return [foundStart, foundEnd];
}

export function binarySearch(nums: number[], find: number): number {
  function bs(start: number, end: number) {
    if (nums[start] === find) return start;
    if (nums[end] === find) return end;
    if (start >= end) return -1;

    const mid = Math.floor((start + end) / 2);

    if (find <= nums[mid]) return bs(start, mid);
    else return bs(mid + 1, end);
  }

  return bs(0, nums.length);
}

File: returned/python/today0/find_first_and_last_position_of_element_in_sorted_array.py
# https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/
from typing import List

class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        def binary_search(searching_left: bool):    
            left = 0
            right = len(nums) - 1
            index = -1
            while left <= right:
                mid = (left + right) // 2
                if target < nums[mid]:
                    right = mid - 1
                elif target > nums[mid]:
                    left = mid + 1
                else:
                    index = mid
                    if searching_left:
                        right = mid - 1
                    else:
                        left = mid + 1    
            return index
        left = binary_search(True)
        right = binary_search(False)
        return [left, right]
